File: appkit-creating-and-customizing-the-touch-bar/Objective-C/NSTouchBar Catalog/TestViewControllers/ColorPickerViewController.m
/*
See the LICENSE.txt file for this sample’s licensing information.

Abstract:
View controller responsible for showing different color picker items.
*/

#import "ColorPickerViewController.h"
#import "SplitViewController.h"
#import "PrimaryViewController.h"

static NSTouchBarItemIdentifier ColorPickerItemIdentifier = @"com.TouchBarCatalog.colorPicker";
static NSTouchBarCustomizationIdentifier ColorPickerCustomizationIdentifier = @"com.TouchBarCatalog.colorPickerViewController";

typedef NS_ENUM(NSInteger, ColorPickerType) {
    ColorPickerTypeColor = 1002,
    ColorPickerTypeText = 1003,
    ColorPickerTypeStroke = 1004
};

@interface ColorPickerViewController () <NSTouchBarDelegate>

@property (strong) NSColorPickerTouchBarItem *colorPickerItem;
@property (strong) IBOutlet NSButton *customColors;

@property ColorPickerType pickerType;

@end


#pragma mark -

@implementation ColorPickerViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    _pickerType = ColorPickerTypeColor;
    
    // Note: If you ever want to show the NSTouchBar instance within this view controller, do this:
    // [self.view.window makeFirstResponder:self.view];
}

- (void)invalidateTouchBar
{
    // Set the first responder status when the user selects a radio button.
    [self.view.window makeFirstResponder:self.view];
    
    // Set to nil so you can call makeTouchBar again to recreate the NSTouchBar instance.
    self.touchBar = nil;
}

- (IBAction)customColorsAction:(id)sender
{
    [self invalidateTouchBar];
}

- (IBAction)choiceAction:(id)sender
{
    _pickerType = ((NSButton *)sender).tag;
    
    [self invalidateTouchBar];
}

#pragma mark NSTouchBarProvider

- (NSTouchBar *)makeTouchBar
{
    NSTouchBar *bar = [[NSTouchBar alloc] init];
    bar.delegate = self;
    
    bar.customizationIdentifier = ColorPickerCustomizationIdentifier;
    
    // Set the default ordering of items.
    bar.defaultItemIdentifiers = @[ColorPickerItemIdentifier, NSTouchBarItemIdentifierOtherItemsProxy];
    
    bar.customizationAllowedItemIdentifiers = @[ColorPickerItemIdentifier];
    
    bar.principalItemIdentifier = ColorPickerItemIdentifier;
    
    return bar;
}

#pragma mark NSTouchBarDelegate

// The system calls this while constructing the NSTouchBar for each NSTouchBarItem you want to create.
- (nullable NSTouchBarItem *)touchBar:(NSTouchBar *)touchBar makeItemForIdentifier:(NSTouchBarItemIdentifier)identifier
{
    if ([identifier isEqualToString:ColorPickerItemIdentifier])
    {
        if (self.pickerType == ColorPickerTypeColor)
        {
            // Create a bar item containing a button with the standard color picker icon that invokes the color picker.
            _colorPickerItem = [NSColorPickerTouchBarItem colorPickerWithIdentifier:ColorPickerItemIdentifier];
            self.colorPickerItem.target = self;
            self.colorPickerItem.action = @selector(colorAction:);
        }
        else if (self.pickerType == ColorPickerTypeText)
        {
            // Create a bar item containing a button with the standard text color picker icon that invokes the color picker. Use this when picking text colors.
            _colorPickerItem = [NSColorPickerTouchBarItem textColorPickerWithIdentifier:ColorPickerItemIdentifier];
            self.colorPickerItem.target = self;
            self.colorPickerItem.action = @selector(colorAction:);
        }
        else if (self.pickerType == ColorPickerTypeStroke)
        {
            // Creates a bar item containing a button with the standard stroke color picker icon that invokes the color picker.
            // Use this when picking stroke colors.
            _colorPickerItem = [NSColorPickerTouchBarItem strokeColorPickerWithIdentifier:ColorPickerItemIdentifier];
            self.colorPickerItem.target = self;
            self.colorPickerItem.action = @selector(colorAction:);
        }
        
        if (self.customColors.state == NSControlStateValueOn)
        {
            // Use a custom color list for the picker.
            self.colorPickerItem.colorList = [[NSColorList alloc] init];
            [self.colorPickerItem.colorList setColor:[NSColor systemRedColor] forKey:@"Red"];
            [self.colorPickerItem.colorList setColor:[NSColor systemGreenColor] forKey:@"Green"];
            [self.colorPickerItem.colorList setColor:[NSColor systemBlueColor] forKey:@"Blue"];
        }
        
        self.colorPickerItem.customizationLabel = NSLocalizedString(@"Color Picker", @"");
        
        return self.colorPickerItem;
    }
    
    return nil;
}

- (void)colorAction:(id)sender
{
    NSLog(@"Color Chosen = %@\n", ((NSColorPickerTouchBarItem *)sender).color);
}

@end

File: appkit-creating-and-customizing-the-touch-bar/Swift/NSTouchBar Catalog/ColorPickerViewController.swift
/*
See the LICENSE.txt file for this sample’s licensing information.

Abstract:
View controller responsible for showing different color picker items.
*/

import Cocoa

class ColorPickerViewController: NSViewController {
    
    let colorPickerBar = NSTouchBar.CustomizationIdentifier("com.TouchBarCatalog.colorPickerBar")
    
    static let color = NSTouchBarItem.Identifier("com.TouchBarCatalog.TouchBarItem.color")
    static let text = NSTouchBarItem.Identifier("com.TouchBarCatalog.TouchBarItem.text")
    static let stroke = NSTouchBarItem.Identifier("com.TouchBarCatalog.TouchBarItem.stroke")
    
    @IBOutlet weak var customColors: NSButton!

    private enum PickerTypeButtonTag: Int {
        case color = 1002, text = 1003, stroke = 1004
    }
    
    private var selectedItemIdentifier: NSTouchBarItem.Identifier = color
    
    // MARK: - NSTouchBarProvider
    
    override func makeTouchBar() -> NSTouchBar? {
        let touchBar = NSTouchBar()
        touchBar.delegate = self
        
        touchBar.customizationIdentifier = colorPickerBar
        touchBar.defaultItemIdentifiers = [selectedItemIdentifier]
        touchBar.customizationAllowedItemIdentifiers = [selectedItemIdentifier]
        touchBar.principalItemIdentifier = selectedItemIdentifier
        
        return touchBar
    }
    
    // MARK: - Action Functions
    
    @IBAction func choiceAction(_ sender: AnyObject) {
        guard let button = sender as? NSButton,
            let choice = PickerTypeButtonTag(rawValue: button.tag) else { return }
        
        switch choice {
        case .color:
            selectedItemIdentifier = ColorPickerViewController.color
            
        case .text:
            selectedItemIdentifier = ColorPickerViewController.text
            
        case .stroke:
            selectedItemIdentifier = ColorPickerViewController.stroke
        }
        
        touchBar = nil
    }
    
    @IBAction func customColorsAction(_ sender: AnyObject) {
        touchBar = nil
    }
    
    @objc
    func colorDidPick(_ colorPicker: NSColorPickerTouchBarItem) {
        print("Picked color: \(colorPicker.color)")
    }
}

// MARK: - NSTouchBarDelegate

extension ColorPickerViewController: NSTouchBarDelegate {
    
    // The system calls this while constructing the NSTouchBar for each NSTouchBarItem you want to create.
    func touchBar(_ touchBar: NSTouchBar, makeItemForIdentifier identifier: NSTouchBarItem.Identifier) -> NSTouchBarItem? {
        let colorPicker: NSColorPickerTouchBarItem
        
        switch identifier {
        case ColorPickerViewController.color:
            colorPicker = NSColorPickerTouchBarItem.colorPicker(withIdentifier: identifier)
            
        case ColorPickerViewController.text:
            colorPicker = NSColorPickerTouchBarItem.textColorPicker(withIdentifier: identifier)
            
        case ColorPickerViewController.stroke:
            colorPicker = NSColorPickerTouchBarItem.strokeColorPicker(withIdentifier: identifier)
            
        default:
            return nil
        }
        
        colorPicker.customizationLabel = NSLocalizedString("Choose Photo", comment: "")
        colorPicker.target = self
        colorPicker.action = #selector(colorDidPick(_:))
        
        if customColors.state == NSControl.StateValue.on {
            let colorList = ["Red": NSColor.systemRed, "Green": NSColor.systemGreen, "Blue": NSColor.systemBlue]
            colorPicker.colorList = NSColorList()
            
            for (key, color) in colorList {
                colorPicker.colorList.setColor(color, forKey: key)
            }
        }
        
        return colorPicker
    }
    
}

